distributed/scheduler.py

```diff
@@ -874,363 +858,265 @@ def _to_dict_no_nest(self, *, exclude: Container[str] = ()) -> dict:
+
+    #: False
+    #:     Each of :attr:`host_restrictions`, :attr:`worker_restrictions` and
+    #:     :attr:`resource_restrictions` is a hard constraint: if no worker is available
+    #:     satisfying those restrictions, the task cannot go into the "processing" state
+    #:     and will instead go into the "no-worker" state.
+    #: True
+    #:     The above restrictions are mere preferences: if no worker is available
+    #:     satisfying those restrictions, the task can still go into the
+    #:     "processing" state and be sent for execution to another connected worker.
+    loose_restrictions: bool
+
+    #: Whether or not this task is an Actor
+    actor: bool
+
+    #: The group of tasks to which this one belongs
+    group: TaskGroup
+
+    #: Same as of group.name
+    group_key: str
+
+    #: Metadata related to task
+    metadata: dict[str, Any]
+
+    #: Task annotations
+    annotations: dict[str, Any]
+
+    #: Cached hash of :attr:`~TaskState.client_key`
+    _hash: int
 
-    __slots__ = (
-        # === General description ===
-        "actor",
-        # Key name
-        "key",
-        # Hash of the key name
-        "_hash",
-        # Key prefix (see key_split())
-        "prefix",
-        # How to run the task (None if pure data)
-        "run_spec",
-        # Alive dependents and dependencies
-        "dependencies",
-        "dependents",
-        # Compute priority
-        "priority",
-        # Restrictions
-        "host_restrictions",
-        "worker_restrictions",  # not WorkerStates but addresses
-        "resource_restrictions",
-        "loose_restrictions",
-        # === Task state ===
-        "_state",
-        # Whether some dependencies were forgotten
-        "has_lost_dependencies",
-        # If in 'waiting' state, which tasks need to complete
-        # before we can run
-        "waiting_on",
-        # If in 'waiting' or 'processing' state, which tasks needs us
-        # to complete before they can run
-        "waiters",
-        # In in 'processing' state, which worker we are processing on
-        "processing_on",
-        # If in 'memory' state, Which workers have us
-        "who_has",
-        # Which clients want us
-        "who_wants",
-        "exception",
-        "exception_text",
-        "traceback",
-        "traceback_text",
-        "erred_on",
-        "exception_blame",
-        "suspicious",
-        "retries",
-        "nbytes",
-        "type",
-        "group_key",
-        "group",
-        "metadata",
-        "annotations",
-    )
+    __slots__ = tuple(__annotations__)  # type: ignore
 
     def __init__(self, key: str, run_spec: object):
         self.key = key
         self._hash = hash(key)
         self.run_spec = run_spec
-        self._state: str = None  # type: ignore
-        self.exception: str = None  # type: ignore
-        self.exception_blame: TaskState = None  # type: ignore
+        self._state = None  # type: ignore
```

Comment:
The cleanup of these non-None attributes forced to None here is left to a future PR